hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/BaseTableServicePlanActionExecutor.java

```diff
@@ -0,0 +1,200 @@
+        LOG.info("Start to fetch incremental partitions for " + type);
+        Pair<Option<HoodieInstant>, Set<String>> lastInstantAndIncrPartitions = getIncrementalPartitions(type);
+        Option<HoodieInstant> lastCompleteTableServiceInstant = lastInstantAndIncrPartitions.getLeft();
+        Set<String> incrementalPartitions = lastInstantAndIncrPartitions.getRight();
+
+        if (incrementalPartitions.isEmpty()) {
+          // handle the case the writer just commits the empty commits continuously
+          // the incremental partition list is empty we just skip the scheduling
+          LOG.info("Incremental partitions are empty. Skip current schedule " + instantTime);
+          return Collections.emptyList();
+        } else if (lastCompleteTableServiceInstant.isPresent()) {
+          LOG.info("Fetched incremental partitions for " + type + ". " + incrementalPartitions + ". Instant " + instantTime);
+          return new ArrayList<>(incrementalPartitions);
+        } else {
+          // Last complete table service commit maybe archived.
+          // fall back to get all partitions.
+          LOG.info("No previous completed table service instant, fall back to get all partitions");
+        }
+      } catch (Exception ex) {
+        LOG.warn("Failed to get incremental partitions", ex);
+      }
+    }
+
+    // get all partitions
+    LOG.info("Start to fetch all partitions for " + type + ". Instant " + instantTime);
+    return FSUtils.getAllPartitionPaths(context, table.getMetaClient().getStorage(),
+        config.getMetadataConfig(), table.getMetaClient().getBasePath());
+  }
+
+  public Pair<Option<HoodieInstant>, Set<String>> getIncrementalPartitions(TableServiceType type) {
+    Pair<Option<HoodieInstant>, List<String>> missingPair = fetchMissingPartitions(type);
+    Option<HoodieInstant> lastCompleteTableServiceInstant = missingPair.getLeft();
+    List<String> missingPartitions = missingPair.getRight();
+
+    String leftBoundary = lastCompleteTableServiceInstant.isPresent()
+        ? missingPair.getLeft().get().requestedTime() : HoodieTimeline.INIT_INSTANT_TS;
+    String rightBoundary = instantTime;
+    // compute [leftBoundary, rightBoundary) as time window
+    HoodieActiveTimeline activeTimeline = table.getActiveTimeline();
+    Set<String> partitionsInCommitMeta = table.getActiveTimeline().filterCompletedInstants().getCommitsTimeline().getInstantsAsStream()
+        .filter(this::filterCommitByTableType).flatMap(instant -> {
+          try {
+            String completionTime = instant.getCompletionTime();
+            if (completionTime.compareTo(leftBoundary) >= 0 && completionTime.compareTo(rightBoundary) < 0) {
+              HoodieCommitMetadata metadata = TimelineUtils.getCommitMetadata(instant, activeTimeline);
+              // ignore all the clustering operation for both mor and cow table
+              if (!metadata.getOperationType().equals(WriteOperationType.CLUSTER)) {
```

Comment:
why ignore cluster?

Reply:
No need to ignore actually. `filterCommitByTableType` is good enough.